### DISCUSSION TITLE
ROLLBACK behavior under autocommit

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#8209] [Usage Questions]
I'm debugging a hang in SQLAlchemy 1.4.36 + postgresql where pg_activity shows:

```
 pid | duration | query | state
-----+----------+-------+-------
2301877 | 00:41:24.107282 | COMMIT                                                                                                    | idle
 2333975 | 00:41:15.248709 | ROLLBACK                                                                                             | idle
 2333976 | 00:41:15.250297 | SELECT obj.field AS obj_field, ...                                                       +| idle in transaction
         |                 | FROM obj                                                                                                                   +|
         |                 | WHERE '2b4ea8e6-fb28-11ec-bfff-aa4c0b6af2e6'::uuid = obj.id                        |
```

Question: Under what scenarios does SQLA itself issue a ROLLBACK? Will a rollback leave the session correctly autocommitted? Might a rollback leave an autobegin session in an actual *db* open transaction (I know SQLA does some virtual transaction begins)? I'm assuming exceptions raised in the commit() process would cause a rollback?

In our codebase we override commit() to cause collected objects to be expunged from an autocommit=True session to a writer session after which they are then moved back. (see postscript)

I'm suspecting a recent "fix" to the codebase introduced an exception that is being triggered during commit, and that instead of the affected session then ending up without an open transaction at the db, it now instead remains in an open transaction post ROLLBACK. This then causes a hang as other greenlets attempt lazy loads or other automated SQLA db interactions.

However, the select looks strange to me. I note that it's happening in what appears to be a session distinct from the rollback (timestamp and 976 vs 975).

The app's activity at the time was that it had just received a network message causing a new object to be created, which is then added to a one-to-many relationship on its owner. Then, owner and obj are added to a session and committed. obj is a subclass and its base has a join table.

The uuid in the query is most likely that of the obj added immediately prior to the hang. It may be unrelated to the rollback, but its partial form (not joined to its base class tables etc) and timestamp leads me to believe it may be related? Though, if so, I'd imagine any post-rollback cleanups would be done in the same session (975)?

Do you have any educated guesses as to what the select is from? I don't think it is SQLA's attempt to check that obj doesn't already exist in the DB during the obj add. This is because post-hang, the database contains said object. Therefore it was added successfully. I've discussed our SQLA usage in other questions, and you consider it non-standard. See postscript below.

Maybe instead, the select is something SQLA does when the object is add()ed back to what we call our "query session" where we keep long lived objects so that their lazy-loads happen in with autocommit. But its partial form seems weird in that case.

Oh, I just realized those timestamps are durations so actually the COMMIT is the oldest item? I'm not sure how to interpret that ... how does a commit hang *first*?

Thank you in advance for any insights.
A.

[Addl context:  I've previously asked about using SQLA in a complex gevent environment etc and without rehashing all that, suffice it to say that the sessions are in autocommit=False except for one session where queries are serialized which is autocommit=True so that lazy fetches don't leave any transaction open. Everything has been working fine until I perturbed the delicate balance with a fix and now this hang.]

**Comment by zzzeek** [comment]
hi - 

autocommit means at least three different things in the 1.x series of SQLAlchemy (in 2.0 it means only one thing).    so would need to know what you mean by "autocommit", there's not enough information for me to answer your question.

if you mean legacy Session autocommit=True, and you are further not calling `begin()` on the Session so that it least sees a transaction boundary, "rollback" is emitted on literally every statement, because every statement outside of the flush process is emitted on a new connection pool checkout.

if you aren't using autocommit=True, which IMO would be the top priority right now, then ROLLBACK is emitted by the session when you call rollback(); when the session is closed using .close(), and connections in use are released to the pool, that will also call rollback on a connection before it's placed back into the pool.    rollback() and close() is also called if you are using the corresponding transaction-scope and session-scope context manager features.

**Comment by aathan** [comment]
Hi. Thanks for taking a look.

I think you're saying our top priority should be to use autocommit=False. See below for why that's difficult.

autcommit=x in my question means `sessionmaker(bind= dinfo.db_engine, expire_on_commit= False, autocommit= False, class_ = OurSession)` and `grep [.]begin()` yields no matches.

By the way, this app caches Session objects for re-use. I think the legacy there was a concern that calling session_maker() is expensive. In your opinion, is that pointless because all the expense is in the connection handling, which is better controlled via configuration of the connection pool instead? 

Do you have a clue for me as to what that select could be? It seems strange because it's selecting only the portion of the object that is in a polymorphic extension table vs all the other joined columns of the object too. I'm trying to figure out which kind of session (see below) of ours, is emitting it: (a) the one autocommit=True session where we add() objects recently written by autocommit=False sessions or (b) one of the autocommit=True write sessions.

The id in the select is certainly for the most recently written object. The hang must be related.

Below, why autocommit=True seems inevitable for a threaded app in the presence of lazy-loads.

A.

----

I can't find our old discussion to refer you to it. In it I described our gevent based app's session usage, which I think you don't approve of but which fundamentally solves the deadlock on lazy-load issue with SQLA in gevent apps.

The usage pattern implied in https://docs.sqlalchemy.org/en/14/orm/session_basics.html is that objects that have lazy-load attributes must remain in a long-lived session for their lifetime in the app. Particularly if the in-memory model is "one copy of each database object"

In turn, that implies that any greenlet that triggers a lazy-load may leave said session in a factually open transaction on the db. This then sets up a deadlock which occurs later when some other greenlet attempts to write to the database from a different session.

Is there any fallacy thus far?

For that reason autobegin seems fundamentally problematic in threaded environments unless all lazy loads or other implicit begin()s are controlled. The latter are key features of SQLA, so autobegin seems to me to mean SQLA is fundamentally problematic in threaded environments.

I would love to re-orient my thinking if it is flawed. This issue has caused too many headaches and I'd love to grok "a better way."

Given the foregoing, the app's usage pattern is for autcommit=True via exactly one "query session" such that this long-lived session may have lazy-loads happen across greenlets without leaving a factually open transaction on the db. Separately, the app uses additional autcommit=False sessions to commit() new and modified objects to the db. These new or modified objects are explicitly collected at their modification/creation points. Then, when ready to commit, they are expunged out of the exactly one autocommit=True query session, added to an available autocommit=False write session, committed, and then expunged and add()ed back to the one query session.

**Reply by zzzeek** [reply]
> Hi. Thanks for taking a look.
> 
> I think you're saying our top priority should be to use autocommit=False. See below for why that's difficult.
> 
> autcommit=x in my question means `sessionmaker(bind= dinfo.db_engine, expire_on_commit= False, autocommit= False, class_ = OurSession)` and `grep [.]begin()` yields no matches.

that's autocommit=False, that's the default.

> 
> By the way, this app caches Session objects for re-use.

there is absolutely no reason to do that and I would never recommend such a thing.

> I think the legacy there was a concern that calling session_maker() is expensive. 

it is not expensive in the slightest, it's literally a single object constructor that assigns the given arguments as state variables.

> In your opinion, is that pointless because all the expense is in the connection handling, which is better controlled via configuration of the connection pool instead?

There's no "expense" in making new sessions period, I'm not really sure what's being referred towards.  the connection pool used by the engine will reduce the expense of establishing new transactions, yes.   certainly if you are using "autocommit=True" without using `begin()`, that particular expense goes way up.


> 
> Do you have a clue for me as to what that select could be? 

as it seems to occur after you've rolled back a session it is likely some state on one of the objects associated with that session being accessed, such as in log messages or something like that, if not something related to the "cached session" architecture.


> It seems strange because it's selecting only the portion of the object that is in a polymorphic extension table vs all the other joined columns of the object too. 

that sounds like a load which happens when you try to access additional attributes on the object that are local to one of those subclass tables.  notes on that are at https://docs.sqlalchemy.org/en/14/orm/inheritance_loading.html#loading-joined-inheritance .

> 
> A.
> 
> I can't find our old discussion to refer you to it. In it I described our gevent based app's session usage, which I think you don't approve of but which fundamentally solves the deadlock on lazy-load issue with SQLA in gevent apps.

gevent works just fine, if not a little bit brittle in its failure modes, with SQLAlchemy without using autocommit=True in pure autocommit mode (noting again, if you are using autocommit=True, but you only use the Session after calling Session.begin(), that's fine; in 1.4 you no longer need to do that since 1.4 now [allows you to call begin() without using autocommit mode](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#session-features-new-autobegin-behavior)  , so if you are definitely calling begin(), you can turn off autocommit=True.   We are doing this whole thing for Openstack, a lot of which runs on eventlet which is basically the same as gevent with less polish.

now if **your** application has a specific deadlock that autocommit=True "fixes", that's not any reason that gevent is not compatible with a longer-running connection that maintains a transaction, that's just something going wrong in your application.    There have been many bugs in SQLAlchemy's implementation as relates to gevent and eventlet, specifically related to cleanup during failure modes, that in the past 7 years have been fixed as I've been working with Openstack.  There were lots of bugs up until about 3 or 4 years ago and it's been crickets now for quite some time, mostly thanks to [this change in SQLAlchemy 1.1](https://docs.sqlalchemy.org/en/14/changelog/migration_11.html#engines-now-invalidate-connections-run-error-handlers-for-baseexception).    


> 
> The usage pattern implied in https://docs.sqlalchemy.org/en/14/orm/session_basics.html is that objects that have lazy-load attributes must remain in a long-lived session for their lifetime in the app. Particularly if the in-memory model is "one copy of each database object"
> 
> In turn, that implies that any greenlet that triggers a lazy-load may leave said session in a factually open transaction on the db. This then sets up a deadlock which occurs later when some other greenlet attempts to write to the database from a different session.

close the Session when the transaction is done, get rid of it. this will detach all objects and lazy loads will raise.    this sounds like the "putting sessions in a cache" thing causing problems.

> 
> Is there any fallacy thus far?
> 
> For that reason autobegin seems fundamentally problematic in threaded environments unless all lazy loads or other implicit begin()s are controlled. The latter are key features of SQLA, so autobegin seems to me to mean SQLA is fundamentally problematic in threaded environments.

So is your hypothesis that of the [1.6 million daily downloads of SQLAlchemy](https://pypistats.org/packages/sqlalchemy), either nobody is using threads or gevent at all, or that for some reason people have decided to stop posting concurrency-related issues?  We do get concurrency related bug reports still for certain areas like [this one](https://github.com/sqlalchemy/sqlalchemy/issues/8098), so I think people are certainly using threads under very high concurrency.   

It sounds like your application has some session usage patterns in it that would need to be reviewed.    overall though, if you can identify in your application when there's a Session for which you no longer want anything loaded against it, call .close() on it, then `del` it.  that will solve that issue.  good luck

**Reply by aathan** [reply]
> So is your hypothesis that of the [1.6 million daily downloads of SQLAlchemy](https://pypistats.org/packages/sqlalchemy), either nobody is using threads or gevent at all

Not at all. All I meant is that given the fact pattern I've observed, it seems like autobegin is logically problematic. So I'm trying to square observed behaviors with the fact you just stated.

Maybe it's an old bug that's been fixed in the *combination* of lazy-load and autobegin and gevent. The fact is that without autocommit=True our lazy-loads seem to leave database transactions open which lead to deadlocks.

That's really all I'm saying. I'll keep digging.

Raising on lazy-load of a detached object seems to mean not using lazy-loading. But, I guess we'll try running the app that way. Possibly its access pattern relies on lazy loading less than we think.

Thanks again.

**Reply by aathan** [reply]
Maybe it comes down to an edge case where an *logical* autobegin ends up with a factual transaction begin at the database, post lazy-load?

If we assume a long lived session that is never closed where all objects live, then is the intent with autocommit=False for that session's connection to be in only a logical, but not factual, begun transaction post lazy-load?

**Reply by zzzeek** [reply]
> > So is your hypothesis that of the [1.6 million daily downloads of SQLAlchemy](https://pypistats.org/packages/sqlalchemy), either nobody is using threads or gevent at all
> 
> Not at all. All I meant is that given the fact pattern I've observed, it seems like autobegin is logically problematic. So I'm trying to square observed behaviors with the fact you just stated.
> 
> Maybe it's an old bug that's been fixed in the _combination_ of lazy-load and autobegin and gevent. The fact is that without autocommit=True our lazy-loads seem to leave database transactions open which lead to deadlocks.

I'd disagree with that phrasing;  lazyloading doesn't "leave" a transaction open, when you have an active Session(), you should consider that object to be a connection-holding, by default transactional object, full stop.  If you haven't .closed() it, it's a database-live object that will re-acquire a new connection, and by default start a new transaction, any time it's asked to return data.   the objects that you load within the scope of a Session are proxies for database rows.   Asking those objects for information means the Session is being asked to return data.      this is essentially no different than a [pep-249 DBAPI connection](https://peps.python.org/pep-0249/). which features the identical default behavior, that is, "autobegin", where it starts a transaction with first statement emitted and it remains open until rollback() or commit() is invoked.

If you do in fact want the connection itself to not actually reference a real transaction within this process, then you can use [DBAPI level autocommit](https://docs.sqlalchemy.org/en/14/core/connections.html#understanding-the-dbapi-level-autocommit-isolation-level), which will keep the ORM Session with a checked out connection until it's closed, but the connection itself wont refer to transactional resources and you won't have deadlocks.

**Reply by zzzeek** [reply]
there was a time that I thought we might have wanted this for openstack, so using the Session I wrote some [very comprehensive docs](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-transaction-isolation) on how to set AUTOCOMMIT isolation level using the ORM Session at every level, so maybe looking into this will allow you to use more modern patterns.

**Reply by CaselIT** [reply]
> All I meant is that given the fact pattern I've observed, it seems like autobegin is logically problematic. So I'm trying to square observed behaviors with the fact you just stated.

I think that what you should be using is dialect level autocommint, that's *not* deprecated, instead of session level autocommit
https://docs.sqlalchemy.org/en/14/core/connections.html#dbapi-autocommit

> For that reason autobegin seems fundamentally problematic in threaded environments 

all the issues you are encountering are in general caused by sharing a single session between multiple threads, a pattern that is in general discouraged. There is no problem is a session remains confined to a single thread.

>I can't find our old discussion to refer you to it. 

you can search all your discussion/issue/pr on github by searching for `author:<your_username`, like so
https://github.com/sqlalchemy/sqlalchemy/discussions?discussions_q=author%3Aaathan+

Maybe this is the one you wanted to refer towards https://github.com/sqlalchemy/sqlalchemy/discussions/7925

**Reply by aathan** [reply]
EDIT: This response was written before I saw @zzzeek's responses (which I address in my next response)

Thanks for finding that! I must have made a typo in my search term.

> dialect autocommit

I'll research this and I'm also building an isolated example which I hope will exhibit the behavior so that it can be more easily studied. This is time consuming, but I should have done it long ago.

>  single session between multiple threads [discouraged pattern]

I understand and I agree it is at the root.
Where we may have some distance to close is whether that issue limits SQLAs utility too much, and whether there may be an easy fix to that limitation under certain scenarios.

- IF a long lived session is "the way" to ensure app-wide "one object per database object" semantics (so that object graphs are shared across greenlets) and object identity (pythonic "is") is useful and scalar member changes are immediately available ... 
- THEN [discouraged pattern] also implies "one object per database object" i.e. "shared object graphs"  cannot be used in threaded environments
- ALTERNATIVELY we may find conditions that make the [discouraged pattern] not problematic

It feels like there's a solution just around the corner (caveat, I have to make further experiments):
- IF I understood correctly that Session autobegin has a *logical* begin boundary that does not need to correlate with a factual database boundary
-THEN the core issue that makes [discouraged pattern] problematic under autocommit=False may be that Session is sometimes opening a factual db transaction when it could be keeping only a logical transaction open

Again, pending further experiments, this may be limited to post lazy load. It's possible I've misunderstood the autocommit=False hang in some fundamental way, so I will isolate the behavior before I say more.

Thanks again for all the help and for SQLAlchemy!
A.

**Reply by aathan** [reply]
I wrote my previous reply not having seen
 @zzzeek 's intervening replies.

> I'd disagree with that phrasing; lazyloading doesn't "leave" a transaction open, when you have an active Session(), you should consider that object to be a connection-holding, by default transactional object, full stop. If you haven't .closed() it, it's a database-live object that will re-acquire a new connection, and by default start a new transaction, any time it's asked to return data. 

Ok. This is a meaningful reorientation regarding Session. We might draw a distinction though, between the fact that a Session could acquire resources at any moment, and it holding those resources open indefinitely after it has completed the instant task. After all, even in a single-threaded environment where there are two sessions, I don't think the intent is for one session to block the other if the operations on that session are idle?

But I guess you're saying that's precisely the intent. If so, the next question is: Does SQLA want to even tacitly support a one-proxy per database row pattern?

>  it starts a transaction with first statement emitted and it remains open until rollback() or commit() is invoked.

Yes. I understand. No argument. However, we can ask whether it is in fact *necessary* for the Session to not commit() or otherwise release resources following the end of the instant read-only unit of work it has completed -- particularly when that work is an implicit lazy-load. The qualitative issue here being that these are not easily visible in code, since the look like dot attribute access and depend on runtime state of the target.

If it did so, it seems the one-proxy-per-db-row pattern would be easily supported without deadlocks, while enjoying lazy-load sugar, with little or no negative effect. <--- cavalier statement by non-expert

> [DBAPI level autocommit]

Sweet! Perhaps this is effectively equivalent to what I"m describing above.

**Reply by CaselIT** [reply]
The behavior of session tries to mimic the dbapi behavior, where in the default case each operation on the db requires a transaction, that stays open until it's committed or rolled back (or the connection is closed, that's equivalent to rollback).

if you don't want to manage the transactional behavior of a session that you are probably looking for DBAPI level autocommit that disables the above dbapi behaviour allowing sql execution without a transaction

**Reply by aathan** [reply]
This is an exciting avenue for me to try. Thanks for all the fish.

**Comment by aathan** [comment]
I meant to write:
> autcommit=x in my question means sessionmaker(bind= dinfo.db_engine, expire_on_commit= False, autocommit= x, class_ = OurSession) and grep [.]begin() yields no matches.
